### DISCUSSION TITLE
Interceptors

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#7460] [Language Ideas]
Hi All, I was recently watching a video about interceptors and I wanted to provide some thoughts. 

1.  I love the concept. 

2.  I think it could be more powerful if it was implemented as an automatic code fix. Similar to how current code fixer identify a "bad" pattern and suggest an alternative, it would be awesome if an interceptor operated exactly the same way except it was automatic.  This would not only enable advanced AOT scenarios but also library-specific performance optimizations.

3.  I think the term "interceptors" is a bit ambiguous. I suggest "source optimizers". 

@MadsTorgersen

**Comment by iam3yal** [comment]
@TonyValenti 

1. I like this feature too but I don't like the way it's designed specifically the notion of "action in a distance" where a source generator can analyze your code and manipulate it without leaving a trace not necessarily to be malicious but this can cause some headaches, it was nicer if there was some notion of a "origin" to it so you can actually inspect it so when something is being replaced under you, you would know exactly what it was and where to look but I guess that this is the reason it's in preview.

2. What do you mean by automatic? isn't big part of this feature is to make SG replace existing APIs calls? so it is automatic in that sense or I don't understand what you're trying to say?

3. In my opinion the name `Interceptors` makes more sense than `Source Optimizer` even though optimization is probably big part of what it might do `Interceptors` seems more general and easier to understand and explain.

**Reply by FaustVX** [reply]
> 1. I like this feature too but I don't like the way it's designed specifically the notion of "action in a distance" where a source generator can analyze your code and manipulate it without leaving a trace not necessarily to be malicious but this can cause some headaches, it was nicer if there was some notion of a "origin" to it so you can actually inspect it so when something is being replaced under you, you would know exactly what it was and where to look but I guess that this is the reason it's in preview.

This is already a thing, way before `interceptor`. *All* analyzer have access to *all* your code to do whatever they wants.
`interceptor` won't change anything in here.

**Reply by yaakov-h** [reply]
Analysers have read-only access.

**Reply by TonyValenti** [reply]
@iam3yal - Source generators actually cannot replace existing api calls. They are additive-only. To leverage a source generator, you must modify your code to explicitly leverage it. This has advantages but it is a (very manual) opt-in experience.

An automatic code fixer would still require opting-in, but I would envision it as being a easy as referencing a new nuget package: Reference "MyPackage.AOT" or "MyPackage.Optimizer" and get all the goodness the library authors imagine. 

The IDE already has an experience for when a source file does not match the code being debugged. I would imagine enhancing this so that, when debugging, you could toggle between the original view of a file and the enhanced/intercepted/optimized version.

**Reply by iam3yal** [reply]
@TonyValenti

> Source generators actually cannot replace existing api calls.

I know that SG are only additive currently **BUT** this is the very thing they are changing with `Interceptors`, this will allow a SG to replace existing call in your code and no it's not manual it is actually automatic, that's the whole point... it analyzes the code and when the code fits existing pattern, then Roslyn fills the position of the call and boom you get another thing that gets called under you as opposed to the thing you're looking at! 

What they need to do is at least color these calls in another color like RED, just kidding but some feedback needs to be provided to the user that this call is going to be replaced and obviously make sure the debugging experience is great, further into the future when it leaves preview they need to provide a syntax that says that the call is **replaceable** by SG so it's opt-in and not the default, maybe through an attribute on the method itself as opposed to callsite.

> To leverage a source generator, you must modify your code to explicitly leverage it.

Yes, that's not true for `Interceptors`, currently the only thing you have to do is enable it and when you do when you add a NuGet package it can modify your source code by analyzing it and then intercept whatever call it wants.
